core/scheduler/service/replay_worker.go

```diff
@@ -28,281 +36,229 @@ type ReplayScheduler interface {
+	if l := len(pendingRuns); l > 0 {
+		startLogicalTime := pendingRuns[0].GetLogicalTime(jobCron)
+		endLogicalTime := pendingRuns[l-1].GetLogicalTime(jobCron)
+		w.logger.Info("[ReplayID: %s] clearing runs with startLogicalTime: %s, endLogicalTime: %s", replayReq.ID(), startLogicalTime, endLogicalTime)
+		if err := w.scheduler.ClearBatch(ctx, replayReq.Tenant(), replayReq.JobName(), startLogicalTime, endLogicalTime); err != nil {
+			w.logger.Error("unable to clear job run for replay with replay_id [%s]: %s", replayReq.ID(), err)
+			return err
+		}
 	}
 
-	updatedReplayMap := identifyUpdatedRunStatus(replayReq.Runs, incomingRuns)
-	updatedRuns := scheduler.JobRunStatusList(replayReq.Runs).MergeWithUpdatedRuns(updatedReplayMap)
-	inProgressRuns := scheduler.JobRunStatusList(updatedRuns).GetSortedRunsByStates([]scheduler.State{scheduler.StateInProgress})
-	failedRuns := scheduler.JobRunStatusList(updatedRuns).GetSortedRunsByStates([]scheduler.State{scheduler.StateFailed})
-
-	var message string
-	state := scheduler.ReplayStateReplayed
-	if len(inProgressRuns) == 0 && len(failedRuns) == 0 {
-		state = scheduler.ReplayStateSuccess
-		w.l.Info("marking replay %s as success", replayReq.Replay.ID().String())
-	} else if len(inProgressRuns) == 0 && len(failedRuns) > 0 {
-		state = scheduler.ReplayStateFailed
-		message = fmt.Sprintf("found %d failed runs.", len(failedRuns))
-		w.l.Info("marking replay %s as failed", replayReq.Replay.ID().String())
+	// create missing runs
+	missingRuns := scheduler.JobRunStatusList(runs).GetSortedRunsByStates([]scheduler.State{scheduler.StateMissing})
+	me := errors.NewMultiError("create runs")
+	for _, run := range missingRuns {
+		logicalTime := run.GetLogicalTime(jobCron)
+		w.logger.Info("[ReplayID: %s] creating a new run with logical time: %s", replayReq.ID(), logicalTime)
+		if err := w.scheduler.CreateRun(ctx, replayReq.Tenant(), replayReq.JobName(), logicalTime, prefixReplayed); err != nil {
+			w.logger.Error("unable to create job run for replay with replay_id [%s]: %s", replayReq.ID(), err)
+			me.Append(err)
+		}
 	}
 
-	w.l.Info("replay with id %s has updated %d runs", replayReq.Replay.ID().String(), len(updatedRuns))
-	if err := w.replayRepo.UpdateReplay(ctx, replayReq.Replay.ID(), state, updatedRuns, message); err != nil {
-		w.l.Error("unable to update replay with replay_id [%s]: %s", replayReq.Replay.ID().String(), err)
-		return err
-	}
-	raiseReplayMetric(replayReq.Replay.Tenant(), replayReq.Replay.JobName(), state)
-	return nil
+	return me.ToErr()
 }
 
-func identifyUpdatedRunStatus(existingJobRuns, incomingJobRuns []*scheduler.JobRunStatus) map[time.Time]scheduler.State {
+// syncStatus syncs existing and incoming runs
+// replay status: created -> in_progress -> [success, failed]
+// replay runs: [missing, pending] -> in_progress -> [success, failed]
+func syncStatus(existingJobRuns, incomingJobRuns []*scheduler.JobRunStatus) []*scheduler.JobRunStatus {
```

Comment:
can we change it (and some other functions, at least the new ones) to be a method of the type worker? it seems like it's not being used anywhere else.

Reply:
updated this